Remove 63-byte name length limit

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should based on *maint*. -->
<!-- New features should based on *master*. -->

## Description

<!-- Describe your changes in detail. -->
* Remove the 63-byte name length limit (requires a new namelist storage specification).
* Fix the bug that prevented more than 127 names.
* Sort the index in the file to enable fast lookups of data chunks.
* Use a hash map to improve name->id mapping performance.
* Always memory map the index, even in the read/write mode. Use of the read/write mode is no longer discouraged.
* Buffer writes of small data chunks.
* Improve consistency of error reporting.
* GSD v2 can read, write, and append to existing v1 files but names are still truncated to 63 chars and read performance will be poor for files with thousands of chunks.
* Provide an upgrade method that performs an in-place conversion of a v1 file into a v2 file.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
HOOMD v3 will allow storing many logged quantities in GSD files within a fully consistent namespace with long names. Thus, GSD needs to support both long names and many data chunks in a frame. This was not possible in the GSD v1 file format specification. This PR introduces a new v2 specification that 1) changes how the names are stored in the file and 2) sorts the index by both frame and id to enable fast lookups of chunks when there are many thousands of chunks in the file. The limit of 65535 unique data chunks is still in place.

While implementing these new features which will bump GSD to v2.0, I also implemented breaking API changes to make error reporting more consistent across all API calls and a number of other performance improvements.

Testing with microbenchmarks (`benchmark-write.cc` and `benchmark-read.cc`), the new GSD can write a 1-double data chunk in 0.4 microseconds (in files with 1-65535 data chunks), and the latency to read a 1-double data chunk from such a file is 0.7 microseconds.

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves: #43, #59

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
Additional unit tests.

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

This PR includes the updated change log.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/gsd/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**GSD Contributor Agreement**](https://github.com/glotzerlab/gsd/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/gsd/blob/master/doc/credits.rst).

b-butler commented on scripts/benchmark-write.cc:
```diff
@@ -0,0 +1,72 @@
+    }
+    fsync(handle.fd);
+
+    auto t1 = chrono::high_resolution_clock::now();
```
Why are you only timing half of the writes?

joaander replied:
To obtain timing for the typical case and exclude the time taken to write the initial namelist and index expansions.

b-butler replied:
I figured something like that, but I was curious.

vyasr commented on gsd/gsd.c:
```diff
@@ -530,199 +447,213 @@ static int gsd_is_entry_valid(struct gsd_handle* handle, size_t idx)
     {
-        return GSD_ERROR_FILE_CORRUPT;
+        return GSD_ERROR_INVALID_ARGUMENT;
     }
```
Do you need a separate branch for memory mapped files here?

joaander replied:
No, the current implementation always reads the name list directly. It needs to fully read on open anyways to hash the names for lookup.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+        gsd_util_zero_memory(buf->data + buf->reserved,
+                             sizeof(struct gsd_index_entry) * (new_reserved - buf->reserved));
+        buf->reserved = new_reserved;
+    }
```
Can you define a `gsd_buffer_resize` function or so that can be reused in other places like the function for appending to a byte buffer?

joaander replied:
I could have, that might have made `gsd_flush_write_buffer` cleaner to implement. At this point, the code is implemented and working so I'm not going to change it.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+    return GSD_SUCCESS;
+}
+
+inline static int gsd_cmp_index_entry(const void* p1, const void* p2)
```
Why can't you just assign types to the function signature?

joaander replied:
I had previously used this function with qsort which required the `void *` signature. Will change it.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+
+    if (a->frame < b->frame)
+    {
+        result = -1;
```
I'd suggest just returning inside each conditional (or changing everything to be if/else clauses) to reduce the number of conditional evaluations) since this function is probably called a lot at a very deep part of the code during any heap operations.

joaander replied:
clang-tidy complains about if clauses after a return. In either case the optimizer is smart and optimizes away the result value. When I looked at this in `perf` output, I saw that the `jmp` instructions for the loop calling `gsd_cmp_index_entry` are placed directly inside the conditionals.

vyasr replied:
OK, I'm always a little wary of trusting compiler optimizations but that's mostly because I haven't looked at the generated assembly enough to see how well they do. I trust your judgement there.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+    }
+
+    // Mac systems deadlock when writing from a mapped region into the tail end of that same region
+    // unmap the index first and copy it over by chunks
```
Is it worth `#ifdef`ing to directly copy on non-Mac systems?

joaander replied:
I don't think so. This method is called rarely and only having 1 implementation to test and debug is a benefit. Changing from the memory mapped read to copying by chunks didn't change performance in any of my measurements.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+
+    // move buffer_index entries to file_index
+    size_t i;
+    for (i = 0; i < handle->buffer_index.size; i++)
```
Just declare `i` in the loop since it's not needed after.

joaander replied:
I would, but some older C compilers throw errors.

vyasr replied:
Why? Is it specifically for `size_t`?

joaander replied:
Declaring the loop variable in a for loop is available starting in C99. GCC 4.8 defaults to the C89 standard, which results in [this compiler error](https://dev.azure.com/glotzerlab/gsd/_build/results?buildId=1302&view=logs&j=f1f5612d-6dbb-5da1-cdfa-f96216bc9d2d&t=7d294d6f-716b-5102-8eae-8aeada81840c&l=22). See [this question on stackoverflow](https://stackoverflow.com/questions/1287863/c-for-loop-int-initial-declaration) for more discussion. 

We cannot drop gcc 4.8 support because the pyPI manylinux infrastructure uses gcc 4.8. See the [test builds](https://dev.azure.com/glotzerlab/gsd/_build/results?buildId=1304&view=results) which all fail.

It is trivial to request C99 mode in CMake in a way that works across platforms and compilers, but I am not aware of such a feature in setuptools.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+        if (retval != GSD_SUCCESS)
+        {
+            return retval;
+        }
```
Is it important to role back changes if an error is encountered, or is it fine to leave the buffer with partially written information?

joaander replied:
If an error occurs here, there will be some data written to the file that has no index entry pointing to it. Rolling back would only save a small amount of file size and not change the correctness of the file. Previously written frames will still be readable and the frame being written to when the error occurs will either have no data chunks or a subset of the data chunks (e.g. if the user hit a file space quota when part way through writing the index for the frame).

There is only so much we can do about consistency without judicious use of `fsync` and the corresponding loss in performance.

vyasr commented on gsd/gsd.c:
```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+    gsd_util_zero_memory(&header, sizeof(header));
+
+    header.magic = GSD_MAGIC_ID;
+    header.gsd_version = gsd_make_version(2, 0);
```
Should we define an enum at the top of this file defining the current version, so that we don't have to update this in multiple places?

joaander replied:
I added it, but many of those locations require an explicit 2.0 check. If we ever add 3.0, the conditions need to handle 1.0, 2.0, and 3.0 cases explicitly and not based on whatever the current spec is.

vyasr replied:
Yes, I realize that some of them are checks specific to a given API. Having the enum does make it easier to not make mistakes when updating, though, since it saves some effort in looking at which instances are specific version checks and which are the current version.

joaander replied:
Agreed.

vyasr commented on gsd/gsd.c:
```diff
@@ -1187,41 +2039,73 @@ gsd_find_chunk(struct gsd_handle* handle, uint64_t frame, const char* name)
+            else
+            {
+                return &(handle->file_index.data[m]);
+            }
```
Is it worth putting this into a `gsd_heap_search` function or so?

joaander replied:
Maybe, though it is only used once in this function. We can certainly factor it out if we ever need to use it in more than one place.

vyasr commented on gsd/pygsd.py:
```diff
@@ -147,7 +144,7 @@ def __init__(self, file):
             raise RuntimeError("Unsupported GSD file version: "
                                + str(self.__file))
-        if self.__header.gsd_version >= (2 << 16):
+        if self.__header.gsd_version >= (3 << 16):
```
I'd recommend exposing the `gsd_make_version` function to Python rather than doing the bitshifts here again.

joaander replied:
`pygsd.py` is for use without the C library.

vyasr replied:
Right, forgot about that.

vyasr commented on gsd/pygsd.py:
```diff
@@ -158,19 +155,16 @@ def __init__(self, file):
-            if len(sname) == 0:
-                break
+        namelist_raw = self.__file.read(
+            self.__header.namelist_allocated_entries * 64)
```
How does this work for 2.0? Won't you read too little if the names are longer?

joaander replied:
In 2.0, the size of the name list is `namelist_allocated_entries * 64` so this reads all of the names. This is confusing, but I kept it that way for consistency with v1 files where ``namelist_allocated_entries` is the *number* of slots for 64-byte names. In v2, names can be any length and are separated by a single NULL character in the buffer. In v1, each element name was NULL padded out to 64 bytes.

vyasr commented on doc/file-layer.rst:
```diff
@@ -35,59 +40,67 @@ Use-cases
-#. Transparent compression - this gets in the way of parallel I/O. Disk space is cheap.
+#. Transparent conversion between float and double: Callers must take care of
+   this.
+#. Transparent compression - this gets in the way of parallel I/O. Disk space is
```
Use a colon instead of a hyphen for consistency.

joaander replied:
Fixed.

vyasr commented on doc/file-layer.rst:
```diff
@@ -35,59 +40,67 @@ Use-cases
 These capabilities are use-cases that GSD does **not** support, by design.
 
-#. Modify data in the file: GSD is designed to capture simulation data, that raw data should not be modifiable.
+#. Modify data in the file: GSD is designed to capture simulation data, that raw
```
..., *but* that raw data...

joaander replied:
Fixed.

vyasr commented on doc/file-layer.rst:
```diff
@@ -35,59 +40,67 @@ Use-cases
+The file layer is implemented in C (*not C++*) with no dependencies to enable
+trivial installation and incorporation into existing projects. A single header
+and C file completely implement the entire file layer. Python based projects
+that need only read access can use :py:mod:`gsd.pygsd`, a pure python gsd reader
```
"that need only read..."

Consistently capitalize "python" (also in the paragraphs below).

joaander replied:
Fixed.

vyasr commented on tests/test_fl.py:
```diff
@@ -513,10 +519,293 @@ def test_chunk_name_limit(tmp_path, open_mode):
-        # A bug in GSD limits files to 128 chunk names:
-        # see https://github.com/glotzerlab/gsd/issues/43
+        # The GSD specification limits to 65535 names:
         with pytest.raises(Exception):
```
To be specific, this raises a RuntimeError, right?

joaander replied:
Yes, fixed.

joaander commented on CMakeLists.txt:
```diff
@@ -14,19 +14,15 @@ include_directories(${PYTHON_INCLUDE_DIR})
-# option(INSTALL_USER "Install into python user site directory (like python setup.py install --user)" OFF)
-# option(INSTALL_SITE "Install into python system site directory (like python setup.py install)" OFF)
+SET(CMAKE_INSTALL_PREFIX ${PYTHON_SYSTEM_SITE} CACHE PATH "Python site installation directory" FORCE)
+message(STATUS "Setting installation site dir: ${CMAKE_INSTALL_PREFIX}")
```
INSTALL_USER was previously commented. It dates back to before `setup.py` in gsd. Now users can use `pip` or `setup.py` to install gsd. 

This changes makes the system site (or venv site) directory the default. However, we don't even document that `make install` exists, so perhaps we should just remove it with a preference for `setup.py`.

vyasr replied:
I'm fine with preferring setup.py. I don't think it's important to support cmake installation beyond the build directory when setup.py works.

joaander commented on CMakeLists.txt:
```diff
@@ -14,19 +14,15 @@ include_directories(${PYTHON_INCLUDE_DIR})
-    message(STATUS "Setting installation site dir: ${CMAKE_INSTALL_PREFIX}")
-ENDIF()
+if (WIN32)
+add_compile_definitions(_CRT_SECURE_NO_WARNINGS)
```
Windows suggests using: `strncpy_s` and `_sopen_s`. I don't see how `_sopen_s` is any better than `_open`.

Clang-tidy suggests using: `memcpy_s`, `memset_s`, and `strncpy_s`. I don't see how passing the size argument twice to a `_s` variant increases security. Most of the places where gsd uses these methods are immediately preceded with a memory allocation or resizes. Those few places where it is not, I am using `strncpy` to prevent copying past the length of the fixed destination buffer. Additionally, the `_s` variants are in Annex K, an *optional* extension to the C11 standard. For portability, we would need to keep the existing non `_s` variant code path and use it when Annex K is not available.

joaander commented on gsd/fl.pyx:
```diff
@@ -848,6 +842,24 @@ cdef class GSDFile:
+        logger.info('upgrading file: ' + self.name)
+
+        with nogil:
+            retval = libgsd.gsd_upgrade(&self.__handle)
```
GSD itself is not multi-threaded. I use `with nogil` throughout GSD when it drops into a pure C call so that multithreaded user scripts do not serialize on GSD (e.g. one thread for reading data from GSD, one thread for processing that data).

vyasr replied:
Does that primarily impact usage of the C API? Is it possible to accomplish something like that from pure Python, or are you considering Cython usage of GSD?

vyasr replied:
Discussed offline: releasing the GIL here potentially allows multiple threads to be reading GSD files simultaneously, or to use some sort of queue to potentially consume chunks of the GSD file (or multiple GSD files) being added to a queue in separate threads from where they're written.

joaander commented on scripts/benchmark-read.cc:
```diff
@@ -0,0 +1,61 @@
+
+#include "gsd.h"
+
+using namespace std;
```
Fixed.

joaander commented on gsd/gsd.h:
```diff
@@ -104,7 +104,7 @@ enum gsd_error
 enum
 {
-    /// Maximum size of a GSD chunk name in memory
+    /// Size of a GSD name in memory
```
Fixed.

In the GSD v1 spec, all names were exactly 64 bytes. To keep some consistency in the 
headers between v1 and v2, I pad the allocated space for the names in v2 out to the nearest 64 bytes. Individual names can range from 2 bytes (1 char + terminator) to as many bytes as fit in the file.

Triple quote comments are read by doxygen when generating the HTML documentation. Many source code editors (i.e. Visual studio code) also read triple quote comments and display the documentation when the user selects a variable

joaander commented on gsd/gsd.h:
```diff
@@ -184,16 +184,86 @@ struct gsd_index_entry
-     An entry in the list of data chunk names
+    A string name paired with an ID. Used for storing sorted name/id mappings in a hash map.
+*/
+struct gsd_name_id_pair
```
It would, but I'll leave it for now.

joaander commented on gsd/gsd.c:
```diff
@@ -45,13 +48,31 @@ enum
 enum
 {
-    GSD_COPY_BUFFER_SIZE = 1024 * 16
+    GSD_NAME_MAP_SIZE = 57557
```
This is a prime number close to 65k, the max number of names that GSD supports. I didn't find any real measurable difference, but conventional wisdom on hash maps seems to be that modding by a prime number can help mix the bits of the hash function. Many implementations of of the C++ std library use prime numbers for their hash map sizes.

joaander commented on gsd/gsd.c:
```diff
@@ -204,284 +216,189 @@ static ssize_t gsd_pread_retry(int fd, void* buf, size_t count, int64_t offset)
+inline static unsigned long gsd_hash_str(const unsigned char* str)
+{
+    unsigned long hash = 5381; // NOLINT
+    int c;
```
`c` is assigned in the `while` condition: `c = *str++` before use. I would have written this differently, but I just copied and pasted this public domain code from a different source of standard string hash methods in C.

joaander commented on gsd/gsd.c:
```diff
@@ -98,20 +119,11 @@ ssize_t pwrite(int fd, const void* buf, size_t count, int64_t offset)
     @param size_to_zero size of the area to zero in bytes
 */
-void gsd_zero_memory(void* d, size_t size_data, size_t size_to_zero)
+static void gsd_util_zero_memory(void* d, size_t size_to_zero)
```
Yes, I had them all inlined at one point but removed that to break up the `perf` output.

joaander commented on gsd/gsd.c:
```diff
@@ -204,284 +216,189 @@ static ssize_t gsd_pread_retry(int fd, void* buf, size_t count, int64_t offset)
+        if (insert_point->next == NULL)
+        {
+            return GSD_ERROR_MEMORY_ALLOCATION_FAILED;
+        }
```
I tested several hash functions with 65k names and a variety of hash map sizes. With map sizes ~= the number of names, I saw at most 4 collisions, on average much less. In typical files with only 100's of keys, I never observed collisions due to the large map size.

With the DJB hash (used currently), I get just slightly better performance than a binary search. There are fewer strcmp calls, but that is mostly offset by the time in the hash function.
I tested the theoretically better FNV and Murmer hash functions, but actually got lower performance. The maximum number of collisions was higher (I don't understand why) and the hash functions themselves took longer to execute. Even Murmer took longer to execute than DJB even though Murmer operates on 4-bytes wordss.

joaander commented on gsd/gsd.c:
```diff
@@ -530,199 +447,213 @@ static int gsd_is_entry_valid(struct gsd_handle* handle, size_t idx)
+
+        // zero the new memory
+        gsd_util_zero_memory(buf->data + buf->reserved,
+                             sizeof(char) * (new_reserved - buf->reserved));
```
Good catch. I changed this to double `new_reserved` in the loop and only `realloc` once and added size to the start point of the zeroing.

joaander commented:
Thanks @vyasr for the thorough review. I have addressed all of your comments. Let me know if you have any additional suggestions or questions for clarification.
